src/ast/visitor.h

```diff
@@ -0,0 +1,325 @@
+  {
+  }
+
+  // Automatically unpacked and dispatch all variant and vector types into the
+  // suitable visitor method, which includes statements and statement lists.
+  // These may be override by subtype explicitly as well.
+  template <typename... Ts>
+  R visit(std::variant<Ts *...> var)
+  {
+    return std::visit([this](const auto value) -> R { return _visit(*value); },
+                      var);
+  }
+  template <typename T>
+  R visit(const std::vector<T> &var)
+  {
+    for (const auto &value : var) {
+      _visit(value);
+    }
+    return _def();
+  }
+  template <typename T>
+  R visit(std::optional<T> &var)
+  {
+    if (var)
+      return _visit(*var);
+    return _def();
+  }
+  template <typename T>
+  R visit(T *ptr)
+  {
+    if (ptr)
+      return _visit(*ptr);
+    return _def();
+  }
+
+  // There are the runtime-type adaptors that are currently required for
+  // Expression and Statement, but can be removed by encoding this type
+  // information into the AST directly.
+  template <typename T, typename... Ts>
+  R tryVisit(Node *node)
+  {
+    if (auto *t = dynamic_cast<T>(node))
+      return _visit(*t);
+    if constexpr (sizeof...(Ts) != 0) {
+      return tryVisit<Ts...>(node);
+    }
+    return _def();
+  }
+  R visit(Expression &expr)
+  {
+    return tryVisit<Integer *,
+                    PositionalParameter *,
+                    String *,
+                    StackMode *,
+                    Identifier *,
+                    Builtin *,
+                    Call *,
+                    Sizeof *,
+                    Offsetof *,
+                    Map *,
+                    Variable *,
+                    Binop *,
+                    Unop *,
+                    FieldAccess *,
+                    ArrayAccess *,
+                    Cast *,
+                    Tuple *,
+                    Ternary *>(&expr);
+  }
+  R visit(Statement &stmt)
+  {
+    return tryVisit<ExprStatement *,
+                    VarDeclStatement *,
+                    AssignMapStatement *,
+                    AssignVarStatement *,
+                    AssignConfigVarStatement *,
+                    Block *,
+                    If *,
+                    Unroll *,
+                    Jump *,
+                    While *,
+                    For *,
+                    Config *>(&stmt);
+  }
+
+  // This is the entrypoint that should be called from the pass itself. It
+  // ensures that it calls the appropriate hooks before walking.
+  R run(Program &program)
+  {
+    return _visit(program);
+  }
+
+private:
+  template <typename T>
+  R _visit(T &t)
+  {
+    // Before dispatching to the specific visit method, we invoke the walk hook
+    // which can be used to enumerate nodes, dump nodes, etc.
+    Impl *impl = static_cast<Impl *>(this);
+    if constexpr (std::is_base_of_v<Node, T>) {
+      Node *node = static_cast<Node *>(&t);
+      impl->preVisit(*node);
+    }
+    return impl->visit(t);
+  }
+  R _def()
```

Comment:
Nit: rename to `_default()`? Thought it was "define" for a bit.

Reply:
I've used `default_value` just to drop the ugly `_` prefix.